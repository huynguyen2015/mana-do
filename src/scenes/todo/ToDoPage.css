#Page__ToDo {
  background: url("/assets/background/todo-bg.jpeg");
  background-repeat: no-repeat,no-repeat;
  background-position: center center,center center;
  background-size: cover,cover;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: flex-start;
}

.ToDo__container {
  width: 500px;
  padding: 24px;
  margin-top: 3rem;
}

.Todo__creation {
  display: flex;
}

.Todo__input {
  flex: 1 1;
}

.ToDo__list {
  display: flex;
  flex-direction: column;
  margin-top: 1.5rem;
  overflow-y: auto;
  height: 400px;
}

.ToDo__item {
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 10px;
}

.ToDo__item > span {
  flex: 1 1;
  text-align: left;
  margin-left: 8px;
}

.Todo__item_complete {
  text-decoration: line-through;
}

.Todo__content {
  flex: 1 1;
}

.Todo__delete {
  outline: none;
  border: none;
  width: 32px;
  height: 32px;
  min-width: auto;
  min-height: auto;
  box-shadow: none;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
}

.Todo__action, .Todo__delete {
  width: 24px;
  height: 24px;
  flex-shrink: 0;
}

.Todo__toolbar {
  display: flex;
  justify-content: space-between;
  margin-top: 24px;
  align-items: center;
}

.Todo__tabs {
  display: flex;
  justify-content: center;
}

.Todo__tabs > *:not(:last-child) {
  margin-right: 8px;
}

.Action__btn.active {}
.Action__btn > span {
  display: none
}
.Action__btn.active > span {
  display: inline-block!important;
  color: var(--primary)
}

.Page__Todo__empty_background {
  flex: 1;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
}

.Page__Todo__empty_background > img {
  height: 250px;
  width: auto;
}
.Page__Todo__empty_background > .description {
  font-size: 11px;
}
