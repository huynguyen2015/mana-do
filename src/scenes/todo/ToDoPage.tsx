import React, {useEffect, useReducer, useRef, useState} from 'react';
import {RouteComponentProps} from 'react-router-dom';
import reducer, {initialState} from '../../store/reducer';
import {
  setTodos,
  createTodo,
  toggleAllTodos,
  deleteAllTodos,
} from '../../store/actions';
import Service from '../../service';
import {TodoStatus} from '../../models/todo';
import {isTodoCompleted} from '../../utils';
import './ToDoPage.css'
import Checkbox from "../../components/Atoms/Checkbox"
import Button from "../../components/Atoms/Button"
import ToDoList from "../../components/ToDoList"

type EnhanceTodoStatus = TodoStatus | 'ALL';

const ToDoPage = ({history}: RouteComponentProps) => {
  const [{todos}, dispatch] = useReducer(reducer, initialState);
  const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
  const inputRef = useRef<HTMLInputElement>(null);

  // useEffect(() => {
  //   (async () => {
  //     const resp = await Service.getTodos();
  //     dispatch(setTodos(resp || []));
  //   })()
  // }, [])

  const onCreateTodo = async (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter' && inputRef.current) {
      try {
        const resp = await Service.createTodo(inputRef.current.value);
        dispatch(createTodo(resp));
        console.log('a1')
        inputRef.current.value = '';
      } catch (e) {
        if (e.response.status === 401) {
          history.push('/')
        }
      }
    }
  }

  const onToggleAllTodo = (e: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(toggleAllTodos(e.target.checked))
  }

  const onDeleteAllTodo = () => {
    dispatch(deleteAllTodos());
  }

  const showTodos = todos.filter((todo) => {
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      default:
        return true;
    }
  });

  const activeTodos = todos.reduce((accum, todo) => {
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0);

  return (
    <div id="Page__ToDo">
      <div className="ToDo__container card">
        <div className="Todo__creation">
          <input
            ref={inputRef}
            className="Todo__input"
            placeholder="What need to be done?"
            onKeyDown={onCreateTodo}
          />
        </div>
        <ToDoList wrapClass="ToDo__list" todos={showTodos} dispatch={dispatch} />
        <div className="Todo__toolbar">
          {todos.length > 0 ?
            <Checkbox checked={activeTodos === 0 && showTodos.length > 0} onChange={onToggleAllTodo}/> : <div/>
          }
          <div className="Todo__tabs">
            <Button className="Action__btn" text="All" onClick={() => setShowing('ALL')}/>
            <Button className="Action__btn" text="Active" onClick={() => setShowing(TodoStatus.ACTIVE)}/>
            <Button className="Action__btn" text="Completed" onClick={() => setShowing(TodoStatus.COMPLETED)}/>
          </div>
          <Button className="Action__btn" text="Clear all todos" onClick={onDeleteAllTodo}/>
        </div>
      </div>
    </div>
  );
};

export default ToDoPage;
