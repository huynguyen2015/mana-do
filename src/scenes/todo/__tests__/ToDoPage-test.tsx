import React from 'react';
import {cleanup, render, screen} from '@testing-library/react';
import ToDoPage from '../ToDoPage';
import { createMemoryHistory, createLocation } from 'history';
import { match } from 'react-router';

const history = createMemoryHistory();
const path = `/route/:id`;
const matchtest: match<{ id: string }> = {
  isExact: false,
  path,
  url: path.replace(':id', '1'),
  params: { id: "1" }
};
const location = createLocation(matchtest.url);

// Note: running cleanup afterEach is done automatically for you in @testing-library/react@9.0.0 or higher
// unmount and cleanup DOM after the test is finished.
afterEach(cleanup);

it('Todo render correct content', () => {
  const todoPage = render(<ToDoPage history={history} location={location} match={matchtest} key={1}/>);
  expect(todoPage).toMatchSnapshot();
});
