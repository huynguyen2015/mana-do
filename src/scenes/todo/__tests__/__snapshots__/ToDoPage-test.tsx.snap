// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo render correct content 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="Page__ToDo"
      >
        <div
          class="ToDo__container card"
        >
          <div
            class="Todo__creation"
          >
            <input
              class="Todo__input"
              placeholder="What need to be done?"
            />
          </div>
          <div
            class="ToDo__list"
          >
            <div
              class="Page__Todo__empty_background"
            >
              <img
                src="/assets/background/empty-bg.svg"
              />
              <div>
                It's seem you don't have urgent task today
              </div>
            </div>
          </div>
          <div
            class="Todo__toolbar"
          >
            <div />
            <div
              class="Todo__tabs"
            >
              <button
                class="Action__btn active"
              >
                <span>
                  ✓
                </span>
                 
                All
              </button>
              <button
                class="Action__btn "
              >
                <span>
                  ✓
                </span>
                 
                Active
              </button>
              <button
                class="Action__btn "
              >
                <span>
                  ✓
                </span>
                 
                Completed
              </button>
            </div>
            <button
              class="Action__btn"
            >
              <span>
                ✓
              </span>
               
              Clear all todos
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      id="Page__ToDo"
    >
      <div
        class="ToDo__container card"
      >
        <div
          class="Todo__creation"
        >
          <input
            class="Todo__input"
            placeholder="What need to be done?"
          />
        </div>
        <div
          class="ToDo__list"
        >
          <div
            class="Page__Todo__empty_background"
          >
            <img
              src="/assets/background/empty-bg.svg"
            />
            <div>
              It's seem you don't have urgent task today
            </div>
          </div>
        </div>
        <div
          class="Todo__toolbar"
        >
          <div />
          <div
            class="Todo__tabs"
          >
            <button
              class="Action__btn active"
            >
              <span>
                ✓
              </span>
               
              All
            </button>
            <button
              class="Action__btn "
            >
              <span>
                ✓
              </span>
               
              Active
            </button>
            <button
              class="Action__btn "
            >
              <span>
                ✓
              </span>
               
              Completed
            </button>
          </div>
          <button
            class="Action__btn"
          >
            <span>
              ✓
            </span>
             
            Clear all todos
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
